### ISSUE TITLE
Usage of collate_fn

### ISSUE THREAD

**Issue by myaldiz** [flairNLP/flair#889]
Hi, 

I noticed when we do the forward pass for sequence_tagger_model, we do the padding and alignment with python for loops right there. However, if we use some collate_fn to do that, DataLoader will handle it parallelly ```if num_workers>0 ```. It may provide significant performance gains.

Thanks.
![ss](https://user-images.githubusercontent.com/11998186/61287985-c035f380-a7f8-11e9-8674-7813bbe9f81b.png)

**Comment by myaldiz** [comment]
In fact, actual forward pass (RNN + LinearPass) compared to the processing tensors and obtaining labels takes nearly no time!
![ss2](https://user-images.githubusercontent.com/11998186/61289009-ed83a100-a7fa-11e9-8d52-a499b25de80e.png)

**Comment by alanakbik** [comment]
Cool, thanks for sharing this. We've been looking for ideas on how to speed everything up. Does tensor manipulation include the embeddings.embed() call? Or is it just taking the embeddings out and padding them?

**Comment by myaldiz** [comment]
> Cool, thanks for sharing this. We've been looking for ideas on how to speed everything up. Does tensor manipulation include the embeddings.embed() call? Or is it just taking the embeddings out and padding them?

I think I included it, lemme check how long embedding the words takes, but I stacked flair forward, backward and glove models. I strongly believe overhead is due to the python for loops.

**Comment by myaldiz** [comment]
Embedding also takes a long time,
![image](https://user-images.githubusercontent.com/11998186/61340298-4c393100-a874-11e9-8916-4b42e620b9d6.png)

**Comment by alanakbik** [comment]
Thanks for sharing this - any help in making this faster is always appreciated. What embeddings did you use here?

**Comment by myaldiz** [comment]
> Thanks for sharing this - any help in making this faster is always appreciated. What embeddings did you use here?

I used FlairPooledForward&Backward+Glove embeddings. 

When we try to move the embedding part inside the data loading part, it gives Cuda initialization error. I did not try the collate_fn but I suspect it will give the same error. 

And the sad thing is we need to embed first, before tensor manipulation part. If we assume the user will not train the language model part, theoretically we gotta be able to embed with multi-threads for speed up, but I am not sure plausibility with PyTorch :)

**Comment by alanakbik** [comment]
Yes that is generally the problem we face in speeding things up. Conceptually, the embedding step is part of the forward pass which is called both during training and during prediction. If we offload too much logic to the data loader (which is outside the model) then our logic becomes distributed because now the data loader is doing things that should be part of the forward pass. 

I was thinking along the lines of using custom multi-processing inside the forward pass to speed up basic tensor operations over lists of Sentences/Tokens, but ideas are still vague at this point.
